### ISSUE TITLE
Can schedule JTs with extra_vars you can't provide on launch

### ISSUE THREAD

**Issue by AlanCoding** [ansible/awx#497]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
 - API

##### SUMMARY
If a JT has no survey and `"ask_variables_on_launch": false`, you can still force it to launch with your own variables by making a schedule that defines those variables inside of `extra_data`.

##### ENVIRONMENT
N/A

##### STEPS TO REPRODUCE
 - Create a JT that does not prompt for variables and has no survey
 - Create a schedule that runs that JT, provide extra variables `foobar: foobaz`

##### EXPECTED RESULTS

1. Validation error saving the schedule
2. Job does not have extra_vars of `foobar: foobaz`

##### ACTUAL RESULTS
no error saving.

job has those extra_vars.

##### ADDITIONAL INFORMATION

Realized while trying to scope out some things around https://github.com/ansible/awx/issues/169

This line of thinking will also lead us down a rabbit hole around what workflow extra variables can mean, artifacts, etc.

**Comment by wenottingham** [comment]
I would assume with #169 we would deep-six this functionality.

**Comment by AlanCoding** [comment]
yes... there will be 4 different places that this applies to, where 2-3 (depending on some extremely specific technicalities) is a similarly pre-existing problem

 - WFJT extra_vars
 - schedule extra_vars
 - WFJT node extra_vars
 - job relaunch with launchtime extra_vars

**Comment by AlanCoding** [comment]
I actually want to solve this along with #478

Reasoning: we have to already do this in some form for job launches.

https://github.com/ansible/awx/blob/641b18fe133d1d82fbfcf223d39431403e1b6743/awx/main/models/jobs.py#L384-L410

We need a little more than what this provides, and that likely involves a separated method for handling the case of extra_vars / extra_data. This all assumes the JT rules for variables, which are now different from the WFJT rules. 478 will make those rules the same. This issue will then re-use the abstraction for extra_vars accept/ignore logic to validate against unusable data given by the user on schedule creation.

The component of 169 that will dwarf the complexity of this is handling encrypted survey answers in extra_data, which makes the logic stateful, not just in regards to the JT/WFJT variables configuration, but also relative to the currently saved data on the configuration object.

Both layers will need to be well-established on the models & in tests before it can be exposed in the API.

**Comment by AlanCoding** [comment]
Other component is for checking validity of survey vars answers.

https://github.com/ansible/awx/blob/641b18fe133d1d82fbfcf223d39431403e1b6743/awx/api/serializers.py#L3333-L3336

And in model

https://github.com/ansible/awx/blob/641b18fe133d1d82fbfcf223d39431403e1b6743/awx/main/models/mixins.py#L230

It's bugging me now how we handle validation differently.

 - If variable fails survey validation, we return an error code
 - if variable is provided, but ask_variables_on_launch is false, we reject it and put it in ignored_fields

This won't really work for the case of schedule creation. You don't want _any_ silent failures, you would clearly prefer to get an error code saving so that you can modify the configuration.

My preferred course-of-action would be to start throwing errors if extra_vars are provided that would be ignored. That's what we do in all other cases like this.

**Comment by AlanCoding** [comment]
My current WIP branch:

https://github.com/ansible/awx/compare/devel...AlanCoding:vars_blocking?expand=1

**Comment by AlanCoding** [comment]
Apparently we didn't do any of this validation for WFJT nodes either. I have added validation logic for this (in addition to the case for schedules) inside #540.

**Comment by AlanCoding** [comment]
Opening again until linked PR is merged.

**Comment by AlanCoding** [comment]
@wenottingham I'm not sure what to do with this.

If we drop the workflow variables when spawning jobs from it (the academically most consistent), that would destroy a ton of existing use patterns. The real problem is that we have no mechanism to signal to the user that they can't provide arbitrary variables to JTs in a workflow.

This is a permissions problem. We require execute_role to add a JT to a workflow.

https://github.com/ansible/awx/blob/cc33109412c55b739e81731b98ae94816510113c/awx/main/access.py#L1694

But that gives workflow admins (who need not be org admins) a unique kind of ability to alter the JT extra_vars arbitrarily by setting the WFJT vars to whatever they want.

This isn't a new problem, but the rest of the saved launch configs feature makes it painful, because we go to such great lengths to respect the `ask_*` variables in other places, and there's a giant gaping hole for WFJT vars, that I honestly don't see anything we can do to fix it.

(note: the title here refers to schedules, that's already taken care of, the API returns a descriptive error message if you provide unallowed variables, it's WFJTs that's the problem)

**Comment by wenottingham** [comment]
What did we allow (or not) in prior releases?

**Comment by AlanCoding** [comment]
Prior behavior was to ignore the job template `ask_variables_on_launch` rule, and apply the workflow variables. I think we started trying to require JT admin permission for adding it to workflows, but that got pushback and it was changed to exec.

I'm fairly sure this is what will be kept moving forward.

This does, for the record, invalidate the launch config (and that creates a little bit more followup work I need to do to check things). So a workflow-spawned job won't be ordinarily relaunchable / schedulable / etc.
